#473 GitHub action to publish python packages to PyPi

hooiv opened the pull request:
# PyPI Publishing Workflow

## Summary
Added a new GitHub Action workflow to publish packages to PyPI using `uv publish`. The workflow only publishes packages with updated versions compared to what's currently on PyPI.

## Implementation
- Created a GitHub Action workflow that can be triggered manually
- Added a Python script to detect packages with updated versions
- Implemented special handling for the monorepo structure with different package locations
- Added dry run support for safe testing before actual publishing
- Supports publishing to both Test PyPI and production PyPI

## Changes
- Added `.github/workflows/publish-to-pypi.yml` workflow file
- Added `scripts/find_updated_packages.py` helper script

## How to Use
1. Go to Actions > Publish to PyPI > Run workflow
2. Configure options (dry run, PyPI URL)
3. Packages with updated versions will be published automatically

Requires the `PYPI_TOKEN` secret to be set in repository settings.

aatmanvaidya commented:
@hooiv can you tell me what all will I need to do to run this on my end?

hooiv commented:
Login/Create account on pypi, verify email , go to 'account settings' , scroll, find 'api tokens' section , generate api token , copy it , go to your github repo settings , in secruity section , select 'secrets and variables'  then select 'Actions', then select 'new repository secret' , set name as "PYPI_TOKEN" , add token/key , click on 'Add secret' , go to vscode , clone your repo , install 'github actions' extension from extensions section in vscode, github actions icon would appear in left side , select it , find workflow sections , 'Publish to PyPI' would be inside , click 'run' or triangle shaped button ,press enter 3 times for dry run .

aatmanvaidya commented on scripts/find_updated_packages.py:
in this file, is it also possible to use git tag's to actually verify if that version exists? because anyone could tamper with the pyroject.toml fie and change its data, in this case git tags become a full proof way to ensure that the tag exists.

hooiv commented:
3b48fdb40ae5875b8cd8608d543ac5631adc6094 commit comment: Creating dedicated artifacts directory to store files that will be uploaded

hooiv commented:
Enhance PyPI workflow with Git Tag Verification and fix Artifact Uploads

This commit introduces two key improvements to the `publish-to-pypi` workflow:

**1. Git Tag Verification (Security Enhancement):**

*   **Purpose:** To prevent unauthorized package publishing by ensuring the version in `pyproject.toml` matches an existing git tag created through the official release process. This mitigates the risk of someone tampering with version numbers.
*   **Implementation:**
    *   The `scripts/find_updated_packages.py` script now includes a `verify_git_tag_exists` function.
    *   Before checking against PyPI, the script verifies if a git tag matching the package name and version (using the configured `tag_format`) exists.
    *   Packages without a matching git tag are skipped, and a warning is logged.
*   **Control:** A new workflow input `verify_tags` (defaulting to `true`) allows enabling/disabling this check.
*   **Reporting:** Issues are logged, and a `tag_verification_issues.txt` file is generated and included in the artifacts.

**2. Artifact Upload Fix:**

*   **Problem:** The workflow previously showed warnings about missing files for artifact uploads, especially during dry runs, as the `dist/` directory might be cleaned or not contain all expected files at the upload step.
*   **Fix:**
    *   A dedicated `artifacts/` directory is now created during the workflow run.
    *   Relevant files (`packages_to_publish.txt`, `tag_verification_issues.txt`) and copies of built package distributions (`dist/*`) are consistently placed into this `artifacts/` directory during the build step.
    *   The `Upload Artifacts` step now uploads the entire `artifacts/` directory.
*   **Result:** Build artifacts and logs are reliably captured and uploaded for every run, including dry runs.

**Files Changed:**

*   `.github/workflows/publish-to-pypi.yml`: Added `verify_tags` input, updated build step to copy artifacts, modified upload step path.
*   `scripts/find_updated_packages.py`: Added git tag verification logic.

aatmanvaidya commented:
hi @hooiv thank you for the changes, I will need some more time to review and test on my end

aatmanvaidya commented on scripts/find_updated_packages.py:
```diff
@@ -0,0 +1,201 @@
+#!/usr/bin/env python3
```
why is a shebang needed here? is it a windows related thing?

aatmanvaidya commented on scripts/find_updated_packages.py:
```diff
@@ -0,0 +1,201 @@
+    check_test_pypi = os.environ.get("CHECK_TEST_PYPI", "true").lower() == "true"
+    
+    # Force tag verification or not (default to true)
+    verify_tags = os.environ.get("VERIFY_TAGS", "true").lower() == "true"
```
don't fetch this value from the `.env`, by default you should be checking for tag verification, I know that this line also force defaults the value to true, but better to refactor to always make sure it happens and not do it using a variable.

aatmanvaidya commented on scripts/find_updated_packages.py:
```diff
@@ -0,0 +1,201 @@
+            print(f"Error checking version for {package_info['name']}: {e}")
+
+    # Write the list of packages to publish to a file
+    with open("packages_to_publish.txt", "w") as f:
+        f.write(",".join(packages_to_publish))
+
+    # Write a report of tag issues if any were found
+    if packages_with_tag_issues:
+        with open("tag_verification_issues.txt", "w") as f:
+            f.write("\n".join(packages_with_tag_issues))
+        print(f"\nWarning: {len(packages_with_tag_issues)} packages have git tag verification issues. See tag_verification_issues.txt")
```
this is also not needed right?
